WooCommerce/Classes/POS/Presentation/Barcode Scanning/HIDBarcodeParser.swift

```diff
@@ -9,27 +22,153 @@ final class HIDBarcodeParser {
     /// Callback that is triggered when a barcode is successfully scanned
     let onScan: (String) -> Void
 
+    private let timerFactory: TimerFactory
+
     private var buffer = ""
+    private var lastKeyPressTime: Date?
+    private var scanTimer: Timer?
 
-    init(configuration: HIDBarcodeParserConfiguration, onScan: @escaping (String) -> Void) {
+    init(configuration: HIDBarcodeParserConfiguration,
+         onScan: @escaping (String) -> Void,
+         timerFactory: TimerFactory = DefaultTimerFactory()) {
         self.configuration = configuration
         self.onScan = onScan
+        self.timerFactory = timerFactory
     }
 
     /// Process a key press event
     /// - Parameter key: The key that was pressed
-    func processKeyPress(_ key: String) {
-        if configuration.terminatingStrings.contains(key) {
-            onScan(buffer)
-            buffer = ""
+    func processKeyPress(_ key: UIKey) {
+        let currentTime = Date()
+
+        // If characters are entered too slowly, it's probably typing and we should ignore it
+        if let lastTime = lastKeyPressTime,
+           currentTime.timeIntervalSince(lastTime) > configuration.maximumInterCharacterTime {
+            resetScan()
+        }
+
+        // Start timing if this is the first key press
+        if scanTimer == nil {
+            startScanTimer()
+        }
+        lastKeyPressTime = currentTime
+
+        let character = key.characters
```

Comment:
Previous use of `charactersIgnoringModifiers` as passed in from the container was convenient, but meant that anything which should be considered with a modifier would just use the raw keystroke, e.g. `2` instead of `@`, which could break some edge case workflows.